boot/bootutil/src/image_validate.c

```diff
@@ -148,81 +169,81 @@ bootutil_img_validate(struct image_header *hdr, const struct flash_area *fap,
     off = hdr->ih_img_size + hdr->ih_hdr_size;
     size = off + hdr->ih_tlv_size;
 
+    /*
+     * Traverse through all of the TLVs, performing any checks we know
+     * and are able to do.
+     */
     for (; off < size; off += sizeof(tlv) + tlv.it_len) {
         rc = flash_area_read(fap, off, &tlv, sizeof tlv);
         if (rc) {
             return rc;
         }
+
         if (tlv.it_type == IMAGE_TLV_SHA256) {
+            /*
+             * Verify the SHA256 image hash.  This must always be
+             * present.
+             */
             if (tlv.it_len != sizeof(hash)) {
                 return -1;
             }
-            sha_off = off + sizeof(tlv);
-        }
-#ifdef MCUBOOT_SIGN_RSA
-        if (tlv.it_type == IMAGE_TLV_RSA2048) {
-            if (tlv.it_len != 256) { /* 2048 bits */
+            rc = flash_area_read(fap, off + sizeof(tlv), buf, sizeof hash);
+            if (rc) {
+                    return rc;
+            }
+            if (memcmp(hash, buf, sizeof(hash))) {
+                    return -1;
+            }
+
+            sha256_valid = 1;
+#ifdef EXPECTED_SIG_TLV
+        } else if (tlv.it_type == IMAGE_TLV_KEYHASH) {
+            /*
+             * Determine which key we should be checking.
+             */
+            if (tlv.it_len > 32) {
```

Comment:
I'm not complaining about this patch since the problem exists elsewhere in the file, but it would be good to replace the magic number 32s throughout with SHA256_HASH_SIZE or so.

Reply:
And actually, shouldn't this be `if (tlv.it_len != 32)`?

Reply:
The original implementation allowed the hash to be less than the full size of the hash itself. You could, for example, only provide 8 bytes of the key hash, which would still be plenty to uniquely identify a hash, as long as the space of keys involved is less than 2^32.